[SPARK-23299][SQL][PYSPARK] Fix __repr__ behaviour for Rows

tbcs opened the pull request:
This is PR is meant to replace #20503, which lay dormant for a while.  The solution in the original PR is still valid, so this is just that patch rebased onto the current master.

Original summary follows.

## What changes were proposed in this pull request?

Fix `__repr__` behaviour for Rows.

Rows `__repr__` assumes data is a string when column name is missing.
Examples,

```
>>> from pyspark.sql.types import Row
>>> Row ("Alice", "11")
<Row(Alice, 11)>

>>> Row (name="Alice", age=11)
Row(age=11, name='Alice')

>>> Row ("Alice", 11)
<snip stack trace>
TypeError: sequence item 1: expected string, int found
```

This is because Row () when called without column names assumes everything is a string.

## How was this patch tested?

Manually tested and a unit test was added to `python/pyspark/sql/tests/test_types.py`.

ashashwat committed:
[SPARK-23299][SQL][PYSPARK] Fix __repr__ behaviour for Rows

Rows __repr__ assumes data is strings when column name is missing.

Examples,
>>> Row ("Alice", "11")
<Row(Alice, 11)>

>>> Row (name="Alice", age=11)
Row(age=11, name='Alice')

>>> Row ("Alice", 11)
<snip stack trace>
TypeError: sequence item 1: expected string, int found

This is because Row () when called without column names assumes
everything is string.

AmplabJenkins commented:
Can one of the admins verify this patch?

tbcs commented:
If this can get merged then it would be great to also backport the patch to the currently supported stable version(s) of Spark (2.4 and 2.3?).  What do I need to do to make that happen?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

felixcheung commented:
@HyukjinKwon

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #104890 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104890/testReport)** for PR 24448 at commit [`8c448dc`](https://github.com/apache/spark/commit/8c448dce858c3c910929823d3d04c9f0cafd1c4e).

HyukjinKwon commented on python/pyspark/sql/types.py:
```diff
@@ -1549,7 +1549,7 @@ def __repr__(self):
                                          for k, v in zip(self.__fields__, tuple(self)))
         else:
-            return "<Row(%s)>" % ", ".join(self)
+            return "<Row(%s)>" % ", ".join("%s" % (fields) for fields in self)
```
https://github.com/apache/spark/pull/20503#discussion_r225846669

tbcs replied:
I applied the suggested change.

HyukjinKwon commented on python/pyspark/sql/tests/test_types.py:
```diff
@@ -739,6 +739,11 @@ def test_timestamp_microsecond(self):
+    # regression test for SPARK-23299
+    def test_row_without_column_name(self):
+        row = Row("Alice", 11)
+        self.assertEqual(repr(row), "<Row(Alice, 11)>")
```
https://github.com/apache/spark/pull/20503#discussion_r225846381

tbcs replied:
>  test non-ascii compatible characters

I have added a test with unicode values.  Of course that breaks in Python 2.7 because %s is used in __repr__.  I think it would be reasonable to change the use of %s to %r for representing the individual tuple values, just as was suggested originally in the JIRA ticket.  I've made that change and adapted the doctest accordingly.  What do you think about this?

HyukjinKwon commented on python/pyspark/sql/tests/test_types.py:
```diff
@@ -739,6 +739,11 @@ def test_timestamp_microsecond(self):
 
+    # regression test for SPARK-23299
+    def test_row_without_column_name(self):
+        row = Row("Alice", 11)
```
https://github.com/apache/spark/pull/20503#discussion_r225844655

tbcs replied:
Indeed, this form of creating Row objects is probably rarely used and should at least be documented.  Maybe it wasn't even intended to be used in the way I did.  Please have a look at the docstring change I've added and let me know if this is appropriate.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10198/
Test PASSed.

SparkQA commented:
**[Test build #104890 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104890/testReport)** for PR 24448 at commit [`8c448dc`](https://github.com/apache/spark/commit/8c448dce858c3c910929823d3d04c9f0cafd1c4e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104890/
Test PASSed.

tbcs committed:
[SPARK-23299][SQL][PYSPARK] Document the use of Row as a tuple value

tbcs committed:
[SPARK-23299][SQL][PYSPARK] Support unicode in Row.__repr__

Change the representation of tuple value Row objects: use repr instead of str
for the individual tuple values for unicode compatibility.

- change Row.__repr__ to use %r instead of %s
- adapt Row's doctest
- add test for unicode tuple values

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10250/
Test PASSed.

SparkQA commented:
**[Test build #104949 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104949/testReport)** for PR 24448 at commit [`ae011c0`](https://github.com/apache/spark/commit/ae011c0217e07702405d8e9a5ca6a53fd3b50626).

SparkQA commented:
**[Test build #104949 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104949/testReport)** for PR 24448 at commit [`ae011c0`](https://github.com/apache/spark/commit/ae011c0217e07702405d8e9a5ca6a53fd3b50626).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104949/
Test PASSed.

holdenk reviewed (approved):
LGTM thanks for the change. For back porting you can open a PR to the branch-2.4 and branch-2.3. We could also backport during the merge, but I think it makes sense to have a discussion around the backporting separately since it may break folks doctests.

holdenk commented:
Merged to master. Thanks for making your first pull request to Spark @tbcs  :) Do you have a JIRA username I can assign the issue to now that it's fixed for tracking?

tbcs commented:
I'll open a PR for the stable branches when I get a chance to do so.

@holdenk: I created a user on Jira: [tbcs](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tbcs)

srowen commented:
I also wouldn't mind back-porting this. I used the varargs Row constructor, which seems normal, and hit this too. But I get that it may break doctests so, hm, unclear.
